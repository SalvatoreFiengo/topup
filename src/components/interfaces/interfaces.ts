import { CountryType, OperatorType } from "./types";

export interface IData {
    countries:CountryType[],
    operators:OperatorType[],
    products:{
        id: string,
        products: string[]
    }[]   
};

export interface IAppMainState {
    phoneNumber: string,
    country: CountryType | null,
    operator: OperatorType | null,
    amount: string | undefined | null,
    cardNumber: string|null
};

export interface IccError {
    isNameError: boolean;
    isCardError: boolean;
    isMonthError: boolean;
    isYearError: boolean;
    isCCVError: boolean;
    isNotCheckout: boolean;
};

export interface ICountriesSelect {
    data:IData;
    setState:(name:string, value:any)=>void;
    country: CountryType | undefined | null;
};

export interface IOperatorsSelect {
    data: IData;
    iso: string|undefined|null;
    setState: (name:string, value:any)=>void;
    operator: OperatorType|null;
};

export interface IPhoneNumberComponent {
    prefix: string;
    setState: (name:string, value:any)=>void;
    phoneNumber: string | undefined;
};

export interface IFormData {  
    fullName: string;
    cardNumber: string;
    month: string;
    year: string;
    CCV: string;
    phoneNumber: string; 
    amount: string; 
    currency: string;  
};

export interface ICreditCardComponent {
    phoneNumberProp: string;
    amountProp: string;
    setState:(name:string, value:any)=>void;
}

export interface IValidation {
    value: string;
    state: IccError
};