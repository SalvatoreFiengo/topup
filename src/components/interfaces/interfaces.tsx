export interface Idata {

    countries:CountryType[],
    operators:OperatorType[],
    products:{
        id: string,
        products: string[]
    }[]
    
};

export type CountryType ={
    iso: string;
    name: string;
    prefix: string;
}
export type OperatorType = {
    id: string,
    iso: string,
    name: string
}
export type State = {
    phoneNumber: string|undefined,
    country: CountryType | null,
    operator: OperatorType | null,
    amount: string | undefined | null
}