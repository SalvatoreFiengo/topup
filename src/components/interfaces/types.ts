export type CountryType ={
    iso: string;
    name: string;
    prefix: string;
}
export type OperatorType = {
    id: string,
    iso: string,
    name: string
}
export type ErrorType ={
    multi: boolean;
}
