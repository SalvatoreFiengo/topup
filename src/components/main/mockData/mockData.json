{
    "countries": [
        {
            "iso": "AF",
            "name": "Afghanistan",
            "prefix": "33"
        },
        {
            "iso": "IE",
            "name": "Ireland",
            "prefix": "18"
        },
        {
            "iso": "DE",
            "name": "Germany",
            "prefix": "25"
        }
    ],
    "operators": [
        {
            "id": "af-internet",
            "iso": "AF",
            "name": "Afghanistan Internet"
        },
        {
            "id": "af-mobile",
            "iso": "AF",
            "name": "Afghanistan Mobile"
        },
        {
            "id": "de-mobile",
            "iso": "DE",
            "name": "Germany Mobile Operator"
        },
        {
            "id": "de-data",
            "iso": "DE",
            "name": "Germany data plans"
        },
        {
            "id": "ie-one",
            "iso": "IE",
            "name": "Ireland Mobile One"
        },
        {
            "id": "ie-two",
            "iso": "IE",
            "name": "Ireland Mobile Two"
        }
    ],
    "products": [
        {
            "id": "af-internet",
            "products": [
                "10 AFN Internet Recharge",
                "20 AFN Internet Recharge",
                "30 AFN Internet Recharge",
                "40 AFN Internet Recharge"
            ]
        },
        {
            "id": "af-mobile",
            "products": [
                "10 AFN Mobile",
                "20 AFN Mobile",
                "30 AFN Mobile",
                "40 AFN Mobile"
            ]
        },
        {
            "id": "de-mobile",
            "products": [
                "10 EUR Mobile",
                "20 EUR Mobile",
                "30 EUR Mobile",
                "40 EUR Mobile"
            ]
        },
        {
            "id": "de-data",
            "products": [
                "10 EUR Data",
                "20 EUR Data",
                "30 EUR Data",
                "40 EUR Data"
            ]
        },
        {
            "id": "ie-one",
            "products": [
                "20 EUR Mobile",
                "50 EUR Mobile",
                "90 EUR Mobile",
                "120 EUR Mobile"
            ]
        },
        {
            "id": "ie-two",
            "products": [
                "10 EUR Mobile",
                "20 EUR Mobile",
                "30 EUR Mobile",
                "40 EUR Mobile"
            ]
        }
    ]
}