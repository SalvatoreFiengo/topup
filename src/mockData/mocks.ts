import { IData } from "../components/interfaces/interfaces";
import { CountryType } from "../components/interfaces/types";

export const countryMock:CountryType = {iso:"IT", name:"Italy", prefix:"+39"};
export const countriesDataMock:IData = {
    countries:[
        {
            iso:"IT", 
            name:"Italy", 
            prefix:"+39"
        }, 
        {
            iso: "IE",
            name: "Ireland",
            prefix: "18"
        }
    ], 
    operators:[], 
    products:[]
};