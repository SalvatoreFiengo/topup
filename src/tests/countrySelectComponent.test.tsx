import React from 'react';
import { shallow, mount } from 'enzyme';
import { IData } from '../components/interfaces/interfaces'; 
import { CountryType } from '../components/interfaces/types';

import CountrySelectComponent from '../components/main/countrySelectComponent';
import { countryMock, countriesDataMock } from '../mockData/mocks';

describe('CountrySelectComponent', ()=>{
    let data: IData;
    let country: CountryType|null;
    let handleSetState: (name: string, value: any) => void;

    test('it renders countrySelectComponent', ()=>{
        const component = shallow(<CountrySelectComponent data={data} setState={handleSetState} country={country}/>);
        expect(component).toMatchSnapshot(); 
    });

    test('it contains "Who would you like to top-up?" and not "Summary" on country=null', ()=>{
        country = null;
        const component = shallow(<CountrySelectComponent data={data} setState={handleSetState} country={country}/>);
        expect(component.text().includes("Who would you like to top-up?")).toBe(true); 
        expect(component.text().includes("Summary")).toBe(false);
    });

    test('it contains "Summary" and not "Who would you like to top-up?" if country=value', ()=>{
        country = countryMock;
        const component = shallow(<CountrySelectComponent data={data} setState={handleSetState} country={country}/>);
        expect(component.text().includes("Summary")).toBe(true);
        expect(component.text().includes("Who would you like to top-up?")).toBe(false);  
    });

    test('it does not renders passed children component (prop.children) if country=null', ()=>{
        country = null;
        const child = <h1>child</h1>
        const component = mount(<CountrySelectComponent data={data} setState={handleSetState} country={country} >{child}</CountrySelectComponent>);
        expect(component.containsMatchingElement(child)).toEqual(false); 
    });

    test('it renders passed children component (prop.children) if country=value', ()=>{
        country = countryMock;
        const child = <h1>child</h1>
        const component = mount(<CountrySelectComponent data={data} setState={handleSetState} country={country} >{child}</CountrySelectComponent>);
        expect(component.containsMatchingElement(child)).toEqual(true); 
    });
 
    test('it renders state "Italy" from data when search is set to "Italy"', ()=>{
        country = null;
        data = countriesDataMock;
        const myState = 'Italy'

        React.useState = jest.fn().mockReturnValue([myState, {}])
        const component = mount(<CountrySelectComponent data={data} setState={handleSetState} country={country} ></CountrySelectComponent>);
        expect(component.text().includes("Italy")).toEqual(true); 
    });
    
    test('it renders state "Ireland" from data when search is set to "Ir"', ()=>{
        country = null;
        data = countriesDataMock;
        const myState = 'Ir'

        React.useState = jest.fn().mockReturnValue([myState, {}])
        const component = mount(<CountrySelectComponent data={data} setState={handleSetState} country={country} ></CountrySelectComponent>);
        expect(component.text().includes("Ireland")).toEqual(true); 
    });
})
