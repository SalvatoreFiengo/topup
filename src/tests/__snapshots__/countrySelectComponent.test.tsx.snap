// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelectComponent it renders countrySelectComponent 1`] = `ShallowWrapper {}`;
